lib/pos_blockchain.go

```diff
@@ -0,0 +1,1331 @@
+package lib
+
+import (
+	"fmt"
+	"math"
+	"time"
+
+	"github.com/deso-protocol/core/collections"
+	"github.com/deso-protocol/core/consensus"
+	"github.com/dgraph-io/badger/v3"
+	"github.com/golang/glog"
+	"github.com/pkg/errors"
+)
+
+// processBlockPoS runs the Fast-Hotstuff block connect and commit rule as follows:
+//  1. Determine if we're missing the parent block of this block.
+//     If so, return the hash of the missing block and add this block to the orphans list.
+//  2. Validate the incoming block, its header, its block height, the leader, and its QCs (vote or timeout)
+//  3. Store the block in the block index and save to DB.
+//  4. try to apply the incoming block as the tip (performing reorgs as necessary). If it can't be applied, we exit here.
+//  5. Run the commit rule - If applicable, flushes the incoming block's grandparent to the DB
+func (bc *Blockchain) processBlockPoS(block *MsgDeSoBlock, currentView uint64, verifySignatures bool) (
+	_success bool, _isOrphan bool, _missingBlockHashes []*BlockHash, _err error) {
+	// If we can't hash the block, we can never store in the block index and we should throw it out immediately.
+	if _, err := block.Hash(); err != nil {
+		return false, false, nil, errors.Wrapf(err, "processBlockPoS: Problem hashing block")
+	}
+	// Get all the blocks between the current block and the committed tip. If the block
+	// is an orphan, then we store it after performing basic validations.
+	// If the block extends from any committed block other than the committed tip,
+	// then we throw it away.
+	lineageFromCommittedTip, err := bc.getLineageFromCommittedTip(block)
+	if err == RuleErrorDoesNotExtendCommittedTip ||
+		err == RuleErrorParentBlockHasViewGreaterOrEqualToChildBlock ||
+		err == RuleErrorParentBlockHeightNotSequentialWithChildBlockHeight ||
+		err == RuleErrorAncestorBlockValidationFailed {
+		// In this case, the block extends a committed block that is NOT the tip
+		// block. We will never accept this block, so mark it as ValidateFailed.
+		if _, innerErr := bc.storeValidateFailedBlockInBlockIndex(block); innerErr != nil {
```

Comment:
I read through this function and I noticed that we are storing a BlockNode whenever we get here, but not storing the block itself. I think this optimization is clever, but it's still dangerous because someone could just send us millions of garbage blocks and cause us to OOM.

Instead, we need a hard and fast rule: We should not store anything in memory OR on disk until we've done some anti-spam validation. I'm still reading through, but my gut would be that we need to do an additional QC check or something harder to pass before we store an invalid block for later.